### PR TITLE
Implement 60 - support for different model output types

### LINKED ISSUE #60: Support for different output types for a model
Right now all MirrorModels are configured to return just a loss. That's fine for training, but for evaluation purposes, we'll need the actual output of the model.

**Acceptance Criteria**:
- MirrorModels can output different kinds of output (text logits, emotion classification, etc)
- The loss is still in a consistent place so it can be found by the trainer
- If at all possible, type safety is preserved (probably using generics)

### PR DESCRIPTION
**Pull request by JansenNye**
Tested using Wikitext dataset.
Closes #60

### CODE REVIEW

File: src/mirror/trainer.py
Comment (Plyb) [inline src/mirror/trainer.py]: Let's make the `MirrorModel`'s have a type argument (a `TypeVar`) here and other places where `MirrorModel` is used (main, callbacks, etc). That will give us better typechecking and allow us to make sure all of the different places can actually handle the shift in return type.

**Review by Plyb — CHANGES_REQUESTED** [review]
`TestCallback` should also have the type parameters added
